test/gurka/test_match.cc

```diff
@@ -124,3 +125,156 @@ D--3--4--C--5--6--E)";
+                            });
+    map.config.put("mjolnir.traffic_extract", "test/data/algorithm_selection/traffic.tar");
+
+    // add live traffic
+    test::build_live_traffic_data(map.config);
+    test::customize_live_traffic_data(map.config, [&](baldr::GraphReader&, baldr::TrafficTile&, int,
+                                                      valhalla::baldr::TrafficSpeed* traffic_speed) {
+      traffic_speed->overall_speed = 50 >> 1;
+      traffic_speed->speed1 = 50 >> 1;
+      traffic_speed->breakpoint1 = 255;
+    });
+
+    test::customize_historical_traffic(map.config, [](DirectedEdge& e) {
+      e.set_constrained_flow_speed(25);
+      e.set_free_flow_speed(75);
+      std::array<float, kBucketsPerWeek> historical;
+      historical.fill(7);
+      for (size_t i = 0; i < historical.size(); ++i) {
+        // TODO: if we are in morning or evening set a different speed and add another test
+      }
+      return historical;
+    });
+
+    // tests below use saturday at 9am and thursday 4am (27 if for leap seconds)
+    size_t second_of_week_constrained = 5 * 24 * 60 * 60 + 9 * 60 * 60 + 27;
+    size_t second_of_week_freeflow = 3 * 24 * 60 * 60 + 4 * 60 * 60 + 27;
+    auto reader = test::make_clean_graphreader(map.config.get_child("mjolnir"));
+    for (auto tile_id : reader->GetTileSet()) {
+      auto tile = reader->GetGraphTile(tile_id);
+      for (const auto& e : tile->GetDirectedEdges()) {
+        current = tile->GetSpeed(&e, baldr::kCurrentFlowMask, 0);
+        EXPECT_EQ(current, 50);
+        historical = tile->GetSpeed(&e, baldr::kPredictedFlowMask, second_of_week_constrained);
+        EXPECT_EQ(historical, 7);
+        constrained = tile->GetSpeed(&e, baldr::kConstrainedFlowMask, second_of_week_constrained);
+        EXPECT_EQ(constrained, 25);
+        freeflow = tile->GetSpeed(&e, baldr::kFreeFlowMask, second_of_week_freeflow);
+        EXPECT_EQ(freeflow, 75);
+      }
+    }
+  }
+};
+
+gurka::map TrafficBasedTest::map = {};
+uint32_t TrafficBasedTest::current = 0, TrafficBasedTest::historical = 0,
+         TrafficBasedTest::constrained = 0, TrafficBasedTest::freeflow = 0;
+
+uint32_t speed_from_edge(const valhalla::Api& api) {
+  uint32_t kmh = -1;
+  const auto& nodes = api.trip().routes(0).legs(0).node();
+  for (int i = 0; i < nodes.size() - 1; ++i) {
+    const auto& node = nodes.Get(i);
+    if (!node.has_edge())
+      break;
+    auto km = node.edge().length_km();
+    auto h = (nodes.Get(i + 1).cost().elapsed_cost().seconds() -
+              node.cost().elapsed_cost().seconds() - node.cost().transition_cost().seconds()) /
+             3600.0;
+    auto new_kmh = static_cast<uint32_t>(km / h + .5);
+    if (kmh != -1)
+      EXPECT_EQ(kmh, new_kmh);
+    kmh = new_kmh;
+  }
+  return kmh;
+}
+
+// map matching currently only allows forward time tracking and invariant so types 0, 1, 3.
+// type 2 arrive_by is not yet supported
```

Comment:
we check all the different types of su pported timedependent map matching as well as all the different sources of speed data